utils/appliance.py

```diff
@@ -2024,3 +2022,40 @@ def provision_appliance_set(appliance_set_data, vm_name_prefix='cfme'):
     logger.info('Done - configuring appliances')
 
     return appliance_set
+
+
+class ApplianceStack(LocalStack):
+
+    def push(self, obj):
+        was_before = self.top
+        super(ApplianceStack, self).push(obj)
+
+        logger.info("Pushed appliance {} on stack (was {} before) ".format(
+            obj.address, getattr(was_before, 'address', 'empty')))
+        if obj.browser_steal:
+            from utils import browser
+            browser.start()
+
+    def pop(self):
+        was_before = super(ApplianceStack, self).pop()
+        current = self.top
+        logger.info(
+            "Popped appliance {} from the stack (now there is {})".format(
+                was_before.address, getattr(current, 'address', 'empty')))
+        if was_before.browser_steal:
+            from utils import browser
+            browser.start()
+        return was_before
+
+stack = ApplianceStack()
+
+
+def get_or_create_current_appliance():
+    if stack.top is None:
+        base_url = conf.env['base_url']
+        if base_url is None or str(base_url.lower()) == 'none':
+            raise ValueError('No IP address specified! Specified: {}'.format(repr(base_url)))
+        stack.push(IPAppliance(urlparse(base_url)))
+    return stack.top
+
+current_appliance = LocalProxy(get_or_create_current_appliance)
```

Comment:
Just a thought (feel free to ignore me if you find this hilarious)
```python
@LocalProxy
def current_appliance():
    # contents of get_or_create_curren_appliance
```

I know, crazy idea, but I guess correct from syntactic PoV :)